### ISSUE TITLE
Question: creating additional restrictions on class properties in the markdown files

### ISSUE THREAD

**Issue by goneall** [spdx/spdx-3-model#413]
In prior discussions we have agreed that profiles can create additional restrictions on properties defined in other profiles (e.g. a `Licensing` profile can make a `declaredLicense` required even if it is optional in the `core` profile).

How would this be expressed in the model markdown files?

**Comment by zvr** [comment]
Is there a class involved?

For example, we already have the case where `Software/File` makes `name` in `Core/Element` mandatory. Take a look at https://github.com/spdx/spdx-3-model/blob/main/model/Software/Classes/File.md

**Comment by goneall** [comment]
@zvr - thanks for the pointer, I was looking for such an example.

There could be a class involved, although there doesn't have to be.  This would be fore the SPDX Lite proposal.

I guess we could take the approach of subclassing each of the Software classes and adding the external restrictions.

It would be nice, however, if we didn't have to do that.  Have a way of saying "if you're using profile X you need the additional restrictions Y".

**Comment by zvr** [comment]
I suppose we could put the corresponding "External properties restrictions" to the profile-level file...

Right now it's being ignored, waiting for https://github.com/spdx/spec-parser/issues/46 -- in which I have written that this might happen, months ago.

We haven't yet thought how these changes will actually work.I mean, they translate to different SHACL constraints based on values of the `profile` attribute. I'm not sure how dynamic can SHACL be... It will probably mean that we have SHACLs per profile, not per class/property.

**Comment by goneall** [comment]
> We haven't yet thought how these changes will actually work.

I'm wrestling with the same thing in the Java library implementation.  The library has a `validate` method that returns a list of validation errors.  I'm adding a list of profiles as a parameter.  In the Java code, I plan on having a map of profile to paramater restrictions generated from the SHACL.

**Comment by goneall** [comment]
> It will probably mean that we have SHACLs per profile, not per class/property.

This is the approach I was thinking of.

**Comment by NorioKobota** [comment]
Thank you for the discussion.
Just an information.
The example of adding the external restrictions is also used in the dataset profile.
https://github.com/spdx/spdx-3-model/blob/main/model/Dataset/Classes/Dataset.md

**Comment by zvr** [comment]
Thank you for commenting, @NorioKobota. The case you are pointing to is equivalent to the [`File`](https://github.com/spdx/spdx-3-model/blob/main/model/Software/Classes/File.md) example I put on my earlier [comment](https://github.com/spdx/spdx-3-model/issues/413#issuecomment-1630477863). This is the case where a class adds restrictions to properties of the ancestor classes (in the case of `Dataset` to properties of `Core/Artifact`, `Software/Package` and `Software/SoftwareArtifact`). This case is well-understood and is being used.

What we have not yet finalized is how to express the same thing when no classes are being involved. Think, for example, the hypothetical case where the `Dataset` profile had the requirement "every `Element` _must_ have a name" (making the property `name` mandatory). Where would this be expressed? My proposal above is to put an  "External properties restrictions" section to the [profile-level file](https://github.com/spdx/spdx-3-model/blob/main/model/Dataset/Dataset.md).

**Comment by NorioKobota** [comment]
Thanks, @zvr.
I overlooked that Dataset class is a subclass of Package class.

> My proposal above is to put an "External properties restrictions" section to the [profile-level file](https://github.com/spdx/spdx-3-model/blob/main/model/Dataset/Dataset.md).

Understood.
In [the previous Lite Profile PR](https://github.com/spdx/spdx-3-model/pull/350/commits/e6e1dfbc9fdc33a70cde976b5af39af7d963dc92), we created a new class to add restriction to the ancestor class, but it would be simpler and better to have external restriction at the profile level.
